File: Coding Problems & My Solutions/Search in Rotated Sorted Array/Search_in_Rotated_Sorted_Array.java
class Solution {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length < 1){
            return -1;
        }
        int start = 0;
        int end = nums.length - 1;
        int mid = 0;
        while (start <= end){
            mid = start + (end - start) / 2;
            if (nums[mid] == target){
                return mid;
            }
            if (nums[start] <= nums[mid]){//left side sorted
                //check if target is in left sorted half
                if (target >= nums[start] && target < nums[mid]){
                    end = mid-1;
                } else {
                    start = mid+1;
                }
            } else {//right side sorted
                //check if target is in right sorted half
                if (target > nums[mid] && target <= nums[end]){
                    start = mid+1;
                } else {
                    end = mid-1;
                }
            }
        }
        return -1;
    }
}

File: Coding Problems & My Solutions/Search in Rotated Sorted Array/Search_in_Rotated_Sorted_Array.swift
class Solution {
    func search(_ nums: [Int], _ target: Int) -> Int {
        var a: Int = 0
        var b: Int = nums.count - 1
        while a <= b {
            let c = a + ((b-a)/2)
            if nums[c] == target {
                return c
            }
            var leftInOrder = nums[a] <= nums[c]
            if leftInOrder {
                if target >= nums[a] && target <= nums[c] {
                    b = c - 1
                } else {
                    a = c + 1
                }
            } else {
                if target >= nums[c] && target <= nums[b] {
                    a = c + 1
                } else {
                    b = c - 1
                }
            }
        }
        return -1
    }
}